src/NServiceBus.Core/OpenTelemetry/Metrics/MetricsFactory.cs

```diff
@@ -0,0 +1,81 @@
+﻿namespace NServiceBus;
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+
+class MetricsFactory(string endpoint, string discriminator) : IMetricsFactory
+{
+    public IMetricsBag MetricsBag() => new MetricBag(endpoint, discriminator);
+}
+
+class MetricBag(string endpoint, string discriminator) : IMetricsBag
+{
+    public IExecutionMetric StartExecution(Histogram<double> meter, params KeyValuePair<string, object>[] tags)
+    {
+        foreach (KeyValuePair<string, object> keyValuePair in tags)
+        {
+            tagList.Add(keyValuePair);
+        }
+
+        var executionMetric = new ExecutionMetric(meter);
+        metrics.Add(executionMetric);
+        return executionMetric;
+    }
+
+    public void Record(params KeyValuePair<string, object>[] tags)
+    {
+        foreach (KeyValuePair<string, object> keyValuePair in tags)
+        {
+            tagList.Add(keyValuePair);
+        }
+        metrics.ForEach(metric => metric.Record(tagList));
+    }
+
+    readonly List<ExecutionMetric> metrics = [];
+    TagList tagList = MeterTags.CommonMessagingMetricTags(endpoint, discriminator);
+}
+
+class ExecutionMetric : IExecutionMetric
```

Comment:
This class currently encapsulates how we record the duration of an execution, and also what tags we then add on success/failure. But do these tags apply to any type of metric of this type? For example, it may make less sense if we want to capture a metric to measure handler construction time. At that time, we don't know yet whether the handle succeeds or fails. I think it's a bit tricky to abstract this here, especially because it's easy to add something in here to the tags, making that apply to a larger set of metrics (assuming we can reuse this for another type of histograms we may want to capture in the future) considering that tags aren't cheap for users.